source/extensions/filters/http/cache/thundering_herd_handler.cc

```diff
@@ -0,0 +1,233 @@
+#include "source/extensions/filters/http/cache/thundering_herd_handler.h"
+
+#include <chrono>
+
+#include "source/common/common/thread_annotations.h"
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+class ThunderingHerdHandlerNone : public ThunderingHerdHandler {
+public:
+  void handleUpstreamRequest(std::weak_ptr<ThunderingHerdRetryInterface>,
+                             Http::StreamDecoderFilterCallbacks* decoder_callbacks, const Key&,
+                             Http::RequestHeaderMap&) override {
+    decoder_callbacks->continueDecoding();
+  }
+  void handleInsertFinished(const Key&, InsertResult) override {}
+};
+
+class ThunderingHerdHandlerBlockUntilCompletion
+    : public ThunderingHerdHandler,
+      public Logger::Loggable<Logger::Id::cache_filter>,
+      public std::enable_shared_from_this<ThunderingHerdHandlerBlockUntilCompletion> {
+  struct Queue {
+    struct Blocked {
+      std::weak_ptr<ThunderingHerdRetryInterface> weak_filter_;
+      Http::StreamDecoderFilterCallbacks* decoder_callbacks_;
+      // dispatcher is captured separately from decoder_callbacks because decoder_callbacks
+      // can be deleted before the dispatcher it returns.
+      Event::Dispatcher& dispatcher_;
+      // request_headers_ may be invalid if weak_filter_ is deleted.
+      Http::RequestHeaderMap& request_headers_;
+    };
+    bool unblockable_{false};
+    // If retry_until_ is in the future, additional requests should do retryFilter as
```

Comment:
I'm not quite following why retry_until_ is needed. Why wouldn't the second lookup find and serve the just-inserted cache entry?